Tree: Undo and repair data design docs (#13251)
microsoft/FluidFramework: packages/dds/tree/docs/undo/README.md

# Undo

This document offers a high-level description of the undo system.
It should be updated once more progress is made on the implementation.

## Abstract vs. Concrete Undo Messages

Conceptually, an undo edit starts as a very abstract and succinct intention:
"Undo changes from prior change _\<revision-tag\>_".

> It ultimately needs to be converted into a more concrete and verbose description of what changes the undo entails
> (e.g., "delete the node at this path").
> This concrete form is the one that the application code (commonly the Forest code) is able to process.

Reviewer: ```suggestion
It ultimately needs to be converted into a concrete description of how the undo changes the current state of the document state
(e.g., "delete the node at this path").
The application (typically through Forest) is able to process this concrete form, represented as a delta.
```

    -   Post-broadcast computation puts the burden of providing historical data on all peers
        (which means that this data must be included in summaries).
-   Access to sequencing information for the undone edit
    -   Pre-broadcast computation may occur before the undone edit is sequenced.
    -   Post-broadcast computation will occur after the undone edit is sequenced.
-   Access to sequencing information for the undo edit
    -   Pre-broadcast computation will occur before the undo edit is sequenced.
    -   Post-broadcast computation will occur after the undo edit is sequenced.
-   Computational load
    -   Pre-broadcast computation only consumes computational resources on the issuing client.
    -   Post-broadcast computation consumes computational resources on all peers.

Note that larger summary sizes and larger message sizes both increase the network/server load and make (down)loading documents slower.

Because the above choice has such a performance impact,
we give applications some agency in choosing the tradeoff that works for them.
We accomplish that by introducing the concept of an "undo window".

## The Undo Window

> The undo window defines how far back, in terms of the edit history,

Author: I expanded on this a bit. Let me know.

> The undo window defines how far back, in terms of the edit history,

Reviewer 2: Looks good!

Note that the issuer of an undo may attempt to undo a change that is within the undo window,
send the adequate undo message,
and find out that by the time the undo message was sequenced,
the edit to be undone had fallen out of the undo window.
In such circumstances, the undo fails (all peers ignore it).
The issuer of the undo can either give up on undoing the change,
or try again in the new context (i.e., undoing an edit outside the undo window).

## Undo Edits Outside The Undo Window

In a low-frequency (i.e. non-live) collaboration environment,
the edit to be undone will commonly lie outside of the undo window.

There are several approaches we could consider:

1. Do the same when the edit to undo is within the undo window but send along the relevant historical data.
2. Compute the net change to the tip state
   (possibly by re-running commands)
   and send that as a normal changeset.
   Such a changeset would be rebased over any concurrent edits.
3. Same as #2 but with instruction for the changeset to be handled differently
   (e.g., by [postbasing](#postbase) it over concurrent edits instead of rebasing it).

## Partial Undo

> Longer term, we may support a more localized undo that only reverts changes within a specific region of the document.

Author: There is a functional difference.
I added an example to clarify, let me know if that helps.

> Longer term, we may support a more localized undo that only reverts changes within a specific region of the document.

Reviewer 2: That helps a lot, thanks!

# Undo

This document offers a high-level description of the undo system.
It should be updated once more progress is made on the implementation.

## Undo Model and Semantics

There are several different choices for what the semantics of undo could be.

> At the same time, there are several different undo-related workflows that consuming application may want to support.

Reviewer: ```suggestion
At the same time, there are several different undo-related workflows that consuming applications may want to support.
```

Understanding those workflows and uncovering what undo semantics would best serve them is still a work in progress,
but two key points have emerged in that space:

> 1. Undo as a feature very important in that it is likely to be the only/primary way that end users access document history.

Reviewer: ```suggestion
1. Undo as a feature is very important in that it is likely to be the only/primary way that end users access document history.
```
